pinot-plugins/pinot-file-system/pinot-adls/src/main/java/org/apache/pinot/plugin/filesystem/AzureGen2PinotFS.java

```diff
@@ -0,0 +1,447 @@
+    try {
+      boolean isDirectory = isDirectory(segmentUri);
+      if (isDirectory && listFiles(segmentUri, false).length > 0 && !forceDelete) {
+        return false;
+      }
+
+      String path = convertUriToAzureStylePath(segmentUri);
+      if (isDirectory) {
+        _fileSystemClient.deleteDirectoryWithResponse(path, true, null, null, Context.NONE).getValue();
+      } else {
+        _fileSystemClient.deleteFile(path);
+      }
+      return true;
+    } catch (DataLakeStorageException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean doMove(URI srcUri, URI dstUri) throws IOException {
+    LOGGER.info("doMove is called with srcUri='{}', dstUri='{}'", srcUri, dstUri);
+    try {
+      // TODO: currently, azure-sdk has a bug in "rename" when the path includes some special characters that gets
+      // changed during the url encoding (e.g '%' -> '%25', ' ' -> '%20')
+      // https://github.com/Azure/azure-sdk-for-java/issues/8761
+//      DataLakeDirectoryClient directoryClient =
+//          _fileSystemClient.getDirectoryClient(convertUriToAzureStylePath(srcUri));
+//      directoryClient.rename(null, convertUriToAzureStylePath(dstUri));
+      copy(srcUri, dstUri);
+      delete(srcUri, true);
+      return true;
+    } catch (DataLakeStorageException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean copy(URI srcUri, URI dstUri) throws IOException {
+    LOGGER.info("copy is called with srcUri='{}', dstUri='{}'", srcUri, dstUri);
+    // If src and dst are the same, do nothing.
+    if (srcUri.equals(dstUri)) {
+      return true;
+    }
+
+    // Remove the destination directory or file
+    if (exists(dstUri)) {
+      delete(dstUri, true);
+    }
+
+    if (!isDirectory(srcUri)) {
+      // If source is a file, we can simply copy the file from src to dst
+      return copySrcToDst(srcUri, dstUri);
+    } else {
+      // In case we are copying a directory, we need to recursively look into the directory and copy all the files and
+      // directories accordingly
+      try {
+        boolean copySucceeded = false;
+        Path srcPath = Paths.get(srcUri.getPath());
+        for (String path : listFiles(srcUri, true)) {
+          // Compute the src path for the given path
+          URI currentSrc =
+              new URI(srcUri.getScheme(), srcUri.getHost(), convertAzureStylePathToUriStylePath(path), null);
+
+          // Compute the destination path for the current path.
+          String relativeSrcPath = srcPath.relativize(Paths.get(convertAzureStylePathToUriStylePath(path))).toString();
+          String newDstPath = Paths.get(dstUri.getPath(), relativeSrcPath).toString();
+          URI newDst = new URI(dstUri.getScheme(), dstUri.getHost(), newDstPath, null);
+
+          if (isDirectory(currentSrc)) {
+            // If src is directory, create one.
+            if (!mkdir(newDst)) {
```

Comment:
According to the interface description of mkdir, it will either return true or throw an exception. Thus, it won't return false, right?

Reply:
changed to `mkdir(newDst)`